lib/claim/token-holder-revenue-fund-contracts-ensemble.js

```diff
@@ -0,0 +1,359 @@
+        _ensemble.set(this, abstractionNames.map(n => ({contract: new TokenHolderRevenueFundContract(walletOrProvider, n)})));
+        _spannedByCurrency.set(this, new Map());
+        _firstAccrualOffset.set(this, firstAccrualOffset > 0 ? firstAccrualOffset : 0);
+    }
+
+    /**
+     * Returns the first accrual's offset index used in the (temporal) span
+     * @returns {number} - First accrual's offset
+     */
+    get firstAccrualOffset() {
+        return _firstAccrualOffset.get(this);
+    }
+
+    /**
+     * Return status of whether contract instance span has been executed
+     * for the given currency
+     * @param {Currency} currency - The currency
+     * @returns {boolean} True if span has been executed, else false
+     */
+    isSpanned(currency) {
+        const spannedByCurrency = _spannedByCurrency.get(this);
+        return spannedByCurrency.get(currencyToKey(currency)) || false;
+    }
+
+    /**
+     * Span (temporally) in accruals and block numbers for the given currency. I.e. infer the
+     * accrual and block number boundaries of the given currency for each contract
+     * instance in the ensemble
+     * @param {Currency} currency - The currency
+     */
+    async span(currency) {
+        const ensemble = _ensemble.get(this);
+        let globalAccrualOffset = _firstAccrualOffset.get(this);
+        const currencyKey = currencyToKey(currency);
+
+        for (const configuration of ensemble) {
+            const closedAccrualsCount = (await configuration.contract.closedAccrualsCount(
+                currency.ct.toString(), currency.id
+            )).toNumber();
+
+            const span = {};
+            span.startAccrual = globalAccrualOffset || 0;
+            span.endAccrual = span.startAccrual + closedAccrualsCount - 1;
+            span.startBlock =
+                (await configuration.contract.closedAccrualsByCurrency(
+                    currency.ct.toString(), currency.id, span.startAccrual
+                )).startBlock.toNumber();
+            span.endBlock =
+                (await configuration.contract.closedAccrualsByCurrency(
+                    currency.ct.toString(), currency.id, span.endAccrual
+                )).endBlock.toNumber();
+
+            if (!configuration.spansByCurrency)
+                configuration.spansByCurrency = new Map();
+
+            configuration.spansByCurrency.set(currencyKey, span);
+
+            globalAccrualOffset = span.endAccrual + 1;
+        }
+
+        _spannedByCurrency.get(this).set(currencyKey, true);
+    }
+
+    /**
+     * Return status of whether the given accrual has been fully claimed
+     * @param {Wallet} wallet - The claimer nahmii wallet
+     * @param {Currency} currency - The currency
+     * @returns {boolean} True if fully claimed, else false
+     */
+    async fullyClaimed(wallet, currency, accrual) {
+        if (!this.isSpanned(currency))
+            await this.span(currency);
```

Comment:
Maybe make `span()` private and call it with `await span.call(this, currency)`?

Reply:
:bow: